### PR TITLE
⚠️ Run webhooks with managers 

### PR DESCRIPTION
**Pull request by fabriziopandini**
**What this PR does / why we need it**:
Given that #3042 for v1alpha4 requires single controllers managing all namespaces, this PR moves the webhooks back in the main manager and run them behind RBAC in case we'll need a client in the future.

**Which issue(s) this PR fixes**:
Fixes https://github.com/kubernetes-sigs/cluster-api/issues/3822

### CODE REVIEW

**Comment by fabriziopandini** [comment]
/hold
while doing some more testing

File: bootstrap/kubeadm/config/default/kustomizeconfig.yaml
Comment (detiber) [inline bootstrap/kubeadm/config/default/kustomizeconfig.yaml:4]: Is this actually needed? I feel like this was something that we previously used to hack together the separate manager and webhook deployments. I'm  not seeing this in the upstream kubuebuilder examples: https://github.com/kubernetes-sigs/kubebuilder/tree/63b35350fe8b4d133c0bdb8fdd9a9a61fe30ac61/testdata/project-v3/config/default

File: bootstrap/kubeadm/config/default/manager_auth_proxy_patch.yaml
Comment (detiber) [inline bootstrap/kubeadm/config/default/manager_auth_proxy_patch.yaml:25]: The upstream kubebuilder example has this, it might be good to leave it just for the sake of similarity even if it's not strictly needed.

File: bootstrap/kubeadm/main.go
Comment (detiber) [inline bootstrap/kubeadm/main.go:105]: Should this be updated to enabled by default and match the current webhook default of port `9443`?

File: bootstrap/kubeadm/main.go
Comment (detiber) [inline bootstrap/kubeadm/main.go:158]: Seems like `setupReconcilers` and `setupWebhooks` will need to be consolidated

**Review by CecileRobertMichon — COMMENTED** [review]
would be good to also describe these changes in https://github.com/kubernetes-sigs/cluster-api/blob/master/docs/book/src/developer/providers/v1alpha3-to-v1alpha4.md

**Comment by fabriziopandini** [comment]
@CecileRobertMichon 

> would be good to also describe these changes in https://github.com/kubernetes-sigs/cluster-api/blob/master/docs/book/src/developer/providers/v1alpha3-to-v1alpha4.md

Done!

File: bootstrap/kubeadm/config/default/kustomizeconfig.yaml
Comment (fabriziopandini) [reply bootstrap/kubeadm/config/default/kustomizeconfig.yaml:4]: This is not in the kubuilder example, but I felt that we are on the safe path if we ensure all the webhook services get a different name for each provider, so during the reshuffle I preserved:

```
secretName: $(SERVICE_NAME)-cert # this secret will not be prefixed since it's not managed by kustomize
```

in `bootstrap/kubeadm/config/certmanager/certificate.yaml` and made the necessary changes for it being resolved in `config/default/kustomize.yaml` which basically consist of creating this file and referencing it

File: bootstrap/kubeadm/config/default/manager_auth_proxy_patch.yaml
Comment (fabriziopandini) [reply bootstrap/kubeadm/config/default/manager_auth_proxy_patch.yaml:25]: If I leave this I get a kustomize error because it is not able to merge args from this patch and from `bootstrap/kubeadm/config/default/manager_webhook_patch.yaml`, so I remove this one because they are not needed

File: bootstrap/kubeadm/main.go
Comment (fabriziopandini) [reply bootstrap/kubeadm/main.go:104]: We are consistent with kubebuilder (webhook disabled by default, but then activated in the manifest via `bootstrap/kubeadm/config/default/manager_webhook_patch.yaml`)
But I can change if there is an agreement.

File: bootstrap/kubeadm/main.go
Comment (fabriziopandini) [reply bootstrap/kubeadm/main.go:156]: I'm personally for keeping things separated because consolidating could make sense here because there are only one reconciler and few webhook, but in CAPI consolidating most likely is not a good idea because the number of reconciler/webhook is much higher.

File: bootstrap/kubeadm/main.go
Comment (vincepri) [reply bootstrap/kubeadm/main.go:104]: Let's enable it by default

File: bootstrap/kubeadm/main.go
Comment (vincepri) [reply bootstrap/kubeadm/main.go:156]: We should remove them, one of the goals of "moving back to a single manager" is to get back in line with kubebuilder project structure and codebase. A familiar codebase is better for newcomers that might already have experience with a Kubebuilder project

File: Makefile
Comment (vincepri) [inline Makefile:313]: Are these `ci` folders not needed any longer?

File: Makefile
Comment (vincepri) [inline Makefile:543]: IIRC we needed this to build the docker container for some old e2e, if those have been removed we can probably safely delete the example provider

File: docs/book/src/developer/providers/v1alpha3-to-v1alpha4.md
Comment (vincepri) [inline docs/book/src/developer/providers/v1alpha3-to-v1alpha4.md:45]:
```suggestion
## Required kustomize changes to have a single manager watching all namespaces and answer to webhook calls
```

File: docs/book/src/developer/providers/v1alpha3-to-v1alpha4.md
Comment (vincepri) [inline docs/book/src/developer/providers/v1alpha3-to-v1alpha4.md:49]: What does this mean?

File: docs/book/src/developer/providers/v1alpha3-to-v1alpha4.md
Comment (vincepri) [inline docs/book/src/developer/providers/v1alpha3-to-v1alpha4.md:46]:
```suggestion
In an effort to simplify the management of Cluster API components, and realign with Kubebuilder configuration, we're requiring some changes to move all webhooks back into a single deployment manager, and to allow Cluster API watch all namespaces it manages.

For a `/config` folder reference, please use the testdata in the Kubebuilder project: https://github.com/kubernetes-sigs/kubebuilder/tree/master/testdata/project-v3/config 
```

File: main.go
Comment (vincepri) [inline main.go:130]:
```suggestion
	fs.IntVar(&webhookPort, "webhook-port", 9443,
```

File: main.go
Comment (vincepri) [inline main.go:131]:
```suggestion
		"Webhook Server port")
```

File: main.go
Comment (vincepri) [reply main.go:130]: This applies to all the other managers as well

File: main.go
Comment (vincepri) [reply main.go:131]: This applies to all the other managers as well

File: main.go
Comment (vincepri) [inline main.go:200]: Let's move this back inline, same goes for the setupWebhooks

Comment (vincepri) [reply main.go:200]: This applies to all the other managers as well

File: test/infrastructure/docker/config/default/manager_auth_proxy_patch.yaml
Comment (vincepri) [inline test/infrastructure/docker/config/default/manager_auth_proxy_patch.yaml:26]: These feature gates shouldn't have been here

File: bootstrap/kubeadm/config/default/manager_auth_proxy_patch.yaml
Comment (vincepri) [reply bootstrap/kubeadm/config/default/manager_auth_proxy_patch.yaml:25]: Given that we're going to enable webhooks by default, we should move all of these args in the main manager file, instead of having them around in places

Comment (vincepri) [reply bootstrap/kubeadm/config/default/manager_auth_proxy_patch.yaml:25]: (Applies to all the other manages as well)

File: Makefile
Comment (fabriziopandini) [reply Makefile:313]: I did not find them used anywhere in our E2E, so...

File: Makefile
Comment (fabriziopandini) [reply Makefile:543]: I will open a good first issue for this

File: docs/book/src/developer/providers/v1alpha3-to-v1alpha4.md
Comment (fabriziopandini) [reply docs/book/src/developer/providers/v1alpha3-to-v1alpha4.md:49]: This instruction assumes the provider's /config folder has the same structure we are maintaining in CAPI (I'm not sure all the providers have the same structure)

File: Makefile
Comment (fabriziopandini) [reply Makefile:543]: https://github.com/kubernetes-sigs/cluster-api/issues/3991

File: main.go
Comment (fabriziopandini) [reply main.go:200]: Leaving this as it is for better code orgazination

File: test/infrastructure/docker/config/default/manager_auth_proxy_patch.yaml
Comment (fabriziopandini) [reply test/infrastructure/docker/config/default/manager_auth_proxy_patch.yaml:26]: fixed

File: bootstrap/kubeadm/config/default/manager_auth_proxy_patch.yaml
Comment (fabriziopandini) [reply bootstrap/kubeadm/config/default/manager_auth_proxy_patch.yaml:25]: done!

File: bootstrap/kubeadm/main.go
Comment (fabriziopandini) [reply bootstrap/kubeadm/main.go]: Done!

File: bootstrap/kubeadm/main.go
Comment (fabriziopandini) [reply bootstrap/kubeadm/main.go:156]: TBH I don't think that inlining everything improves readability for newcomers
However, if we want to tackle this lets this refactor let get this in another PR given that this one is already too big

**Comment by vincepri** [comment]
LGTM pending squash

**Comment by fabriziopandini** [comment]
@vincepri squashed!

**Review by vincepri — APPROVED** [review]
/lgtm
/assign @CecileRobertMichon @detiber  
for approval

**Review by CecileRobertMichon — COMMENTED** [review]
lgtm, thanks for adding the detailed docs!

will leave it up for a minute to give others a chance to review

**Comment by vincepri** [comment]
LGTM pending rebase

**Comment by fabriziopandini** [comment]
@vincepri @detiber @CecileRobertMichon any objection to get this moving & lift the hold?

**Comment by vincepri** [comment]
/lgtm
@CecileRobertMichon 
for approva

**Comment by vincepri** [comment]
@fabriziopandini Not sure if I can merge this like-this, unless we have an immediate patch for the e2e to pass, it seems with these changes they're failing consistently and we should investigate before merging

**Comment by wfernandes** [comment]
Reviewing 👀

**Comment by wfernandes** [comment]
Ran through most of the make commands and manually deployed a CAPD workload cluster using the generated manifests. 
/lgtm

**Comment by fabriziopandini** [comment]
@vincepri the error does not seem related to this specific change, but I will take a look

**Comment by fabriziopandini** [comment]
@vincepri the errors seem to be consistently in the machinepools stuff, I got https://github.com/kubernetes-sigs/cluster-api/pull/3921 out for this

**Comment by fabriziopandini** [comment]
rebase & fixed error on machine pools test
@vincepri this PR should be ready to go now

**Review by vincepri — APPROVED** [review]
/lgtm
/assign @detiber @CecileRobertMichon

File: docs/book/src/developer/providers/v1alpha3-to-v1alpha4.md
Comment (CecileRobertMichon) [reply docs/book/src/developer/providers/v1alpha3-to-v1alpha4.md:53]: should this say as the /config folder in v1alpha3?

File: docs/book/src/developer/providers/v1alpha3-to-v1alpha4.md
Comment (CecileRobertMichon) [inline docs/book/src/developer/providers/v1alpha3-to-v1alpha4.md:65]:
```suggestion
    - Copy the `vars` list into a temporary file to be used later in the process  
```

File: docs/book/src/developer/providers/v1alpha3-to-v1alpha4.md
Comment (CecileRobertMichon) [inline docs/book/src/developer/providers/v1alpha3-to-v1alpha4.md:84]:
```suggestion
    - Verify that fetaure flags required by your container are properly set 
```

File: bootstrap/kubeadm/main.go
Comment (CecileRobertMichon) [reply bootstrap/kubeadm/main.go:162]: can we file an issue for the refactor?

**Review by CecileRobertMichon — COMMENTED** [review]
lgtm overall

sorry for not getting to this earlier @fabriziopandini

**Comment by fabriziopandini** [comment]
Rebased + adressed @CecileRobertMichon comments
/test pull-cluster-api-e2e-main

File: bootstrap/kubeadm/main.go
Comment (fabriziopandini) [reply bootstrap/kubeadm/main.go:162]: https://github.com/kubernetes-sigs/cluster-api/issues/4136

**Comment by vincepri** [comment]
/test pull-cluster-api-e2e-full-main

The panic should be resolved in 1.21

**Review by vincepri — APPROVED** [review]
/lgtm

**Pull request merged**
